### ISSUE TITLE
Official docker image need more modules

### ISSUE THREAD

**Issue by tinuva** [jpmens/mqttwarn#344]
I have been following issue [341](https://github.com/jpmens/mqttwarn/issues/341) and is really excited to be able to use it, however I think it needs some of the prerequisites added to the Dockerfile.

For example the following log entries work on my standalone install:
```
mqttwarn        | 2018-10-30 11:19:16,856 DEBUG [slack] *** MODULE=services/slack.py: service=slack, target=openhabalarm
mqttwarn        | 2018-10-30 11:19:16,857 ERROR [slack] slacker module missing
mqttwarn        | 2018-10-30 11:19:16,858 WARNING [mqttwarn] Notification of slack for '/openhab/notification/alarm' FAILED or TIMED OUT
mqttwarn        | 2018-10-30 11:19:16,859 DEBUG [mqttwarn] Job queue has 5 items to process
mqttwarn        | 2018-10-30 11:19:16,859 DEBUG [mqttwarn] Processor #0 is handling: 'pushbullet' for david
mqttwarn        | 2018-10-30 11:19:16,864 DEBUG [pushbullet] *** MODULE=services/pushbullet.py: service=pushbullet, target=david
mqttwarn        | 2018-10-30 11:19:16,865 WARNING [pushbullet] pushbullet is not installed
mqttwarn        | 2018-10-30 11:19:16,866 WARNING [mqttwarn] Notification of pushbullet for '/openhab/notification/alarm' FAILED or TIMED OUT
````

On the normal install I am able to install the dependencies, but the docker image mean I need to build my own for my specific needs unless you could add some of the popular ones at least.

Many thanks in advance!

**Comment by jpmens** [comment]
We cannot and will not add all prerequisites for all 60+ modules to the Docker image, and "popular" would appear to be in the eyes of the beholder, no?

**Comment by rgitzel** [comment]
@jpmens  Actually, why _not_ add them all?  At worst it makes the Docker build time slightly longer, and the image slightly bigger. Any that you don't use yourself just sit there unused. And they don't interfere with your local Python environment.

It might just mean a bit of playing whack-a-mole explicitly listing them all.

**Comment by tinuva** [comment]
> We cannot and will not add all prerequisites for all 60+ modules to the Docker image, and "popular" would appear to be in the eyes of the beholder, no?

Which ones will work currently? 

I think email works, but if that is the only one, I am not quite sure what the use of this docker image will be.

Please consider adding all if not most.

**Comment by amotl** [comment]
Hi there,

as #127 will wrap the dependency management into the canonical `setup.py` already (at least it will **list** all possible requirements correlated to each service plugin), i am also leaning towards having an easy way to perform a "full" setup in some way or another in the future. This flavor could be spawned from the pip level as an "extra" requirement, acting as a driver for the pip-in-docker building we will approach eventually.

On behalf on dockerized setups, we could think about publishing a distinct `jpmens/mqttwarn-full` image giving all users maximum comfort, no? Otherwise, we would have to think about having popularity shootouts between different service plugins and might end up building `jpmens/mqttwarn-slack`, `jpmens/mqttwarn-xmpp`, but not `jpmens/mqttwarn-pushover` etc. This would explode the matrix eventually.

On the other hand, i am also keen on adding a subcommand to _mqttwarn_ like `mqttwarn activate <plugin>`, which could well be executed interactively on behalf of the user using `docker run` as well.

By combining both tactics, we could get the best out of different worlds, but that's really just my personal two cents.

With kind regards,
Andreas.

**Comment by rgitzel** [comment]
@amotl what is the reason to NOT install all the possible dependencies at once? (Other than taking up disk space.)  I'm not a Python dev, so I don't know if it's a compatibility issue or such.

**Comment by rgitzel** [comment]
@tinuva here's what is currently included: https://github.com/jpmens/mqttwarn/blob/master/Dockerfile#L5

**Comment by jpmens** [comment]
*shrugs*

I will accept a PR for the Dockerfile to include more services; I don't think we should over-complicate this and I also don't think building `jpmens/mqttwarn-*.*` is going to do anybody any good. :-)

**Comment by rgitzel** [comment]
@amotl I'd like to not confuse the use of Docker now with the use of Docker with the `dev` branch, as they are built differently.

But I do already have a candidate, which in fact is the one I've been running for the past few months:  https://github.com/rgitzel/mqttwarn/blob/docker-pip/Dockerfile

**Comment by amotl** [comment]
> what is the reason to NOT install all the possible dependencies at once? (Other than taking up disk space.) I'm not a Python dev, so I don't know if it's a compatibility issue or such.

@rgitzel: Not **all** service pluings have well defined upstream dependencies on PyPI. We were having our times with XMPP for example.

While i definitively want to **approach** a wash & go full setup, i am as conservative as @jpmens here until we got everything rock solid as we probably might come across thinking about whether to publish own stable forks of specific service plugin dependencies to PyPI. It's just a story of turtles all the way down and some amount of tedious homework. I don't want to push this that fast but i am confident everything will converge at some time in the future.

_mqttwarn-ng_ from #127 will include the most important subset of dependencies anyway. For example, every service plugin calling out the HTTP APIs using the fine `requests` package will be covered already.

Don't get me wrong: There **should** be some convenient way to perform a "full" setup, which might also help an expanding #332 when running on CI through `tox` or similar beasts. Until then, i am totally with @jpmens that building `jpmens/mqttwarn-*.*` too early is not going to do anybody any good, but will put more burden on our scarce maintenance resources.

**Comment by egadgetjnr** [comment]
Hey Guys,
Just my 2 cents on this one being somewhat an experienced docker user. 

I think going down the path of having ALL the modules installed on an image would be somewhat troublesome as it would increase the image size exponentially. 

Also, Using a tag to define the modules would also be difficult if there is more than one module requirement. How would you be able to use Pushbullet and MySQL and hipchat all in the one image?

@amotl is on the money though with having a command that you can run to install them, as you’d be able to use your own dockerfile to specify the modules you want if need be. Failing that, could you use arguments to install the plugins on first run? That way when you use ‘docker run’ you will be able to pass arguments for the module installation at the end of the command.
Both of these approaches are what other developers use for their docker repos and it allows full customisation of different scenarios.

**Comment by rgitzel** [comment]
I did an experiment to create a Docker image which includes as many dependencies as I could find reading through the README: https://github.com/rgitzel/mqttwarn/blob/docker-dependencies/Dockerfile#L5

With all those dependencies added, the Docker image is [**420Mb**](https://hub.docker.com/r/rgitzel/mqttwarn-full/tags/)

With [very few dependencies](https://github.com/jpmens/mqttwarn/blob/master/Dockerfile#L6), the "official" image built yesterday is [**356Mb**](https://hub.docker.com/r/jpmens/mqttwarn/tags/)

So basically installing all the extra libraries results in no appreciable difference in size.

Then I modified it [to use Alpine as the base OS](https://github.com/rgitzel/mqttwarn/blob/docker-dependencies-alpine/Dockerfile#L6), which reduced the image to [**117Mb**](https://hub.docker.com/r/rgitzel/mqttwarn-full-alpine/tags/)

That's indeed smaller, but frankly I don't think it's worth the work sorting out all the picky library bits that are missing from Alpine (indeed, the Azure library would be painful).  420Mb isn't anything out of the ordinary, particularly for a server application.

**CAVEAT:**  I haven't tested either of these images beyond running them on the command-line:
```
$ docker run -t rgitzel/mqttwarn-full
Cannot open configuration at /opt/mqttwarn/conf/mqttwarn.ini: [Errno 2] No such file or directory: '/opt/mqttwarn/conf/mqttwarn.ini'
```

Because Python is interpreted, the app probably won't fail for a given plugin until it's run.  But at least those many dependencies _installed_ without any apparent conflict.

@tinuva  could you give one or both of the new images a try?  `rgitzel/mqttwarn-full-alpine` and `rgitzel/mqttwarn-full`.  They hopefully have everything installed that you need.

And if anyone notices any dependency I missed (other than the Mac-dependent ones), let me know and I'll add them.

**Comment by tinuva** [comment]
@rgitze tested both and both work for me at least. Which is smtp+pushbullet+slack 
The test was simple, start the image with my current configuration and then send a test message to a channel that makes use of all my mqttwarn service providers.

Going to leave it on the Alpine image, which is what I prefer where possible.

My docker-compose snippet:
```
  mqttwarn:
    container_name: mqttwarn
    image: rgitzel/mqttwarn-full-alpine
    restart: always
    environment:
      MQTTWARNLOG: stream://sys.stderr
    volumes:
      - ${USERDIR}/docker/mqttwarn:/opt/mqttwarn/conf
```

**Comment by tinuva** [comment]
I think @rgitzel 's experiment is doing well. See the most popular mqttwarn container on hub.docker.com....
https://hub.docker.com/search?q=mqttwarn&type=image

That said, I will be perfectly happy with a base mqttwarn image that is able to install additional dependancies on demand.

**Comment by rgitzel** [comment]
@tinuva Holy smokes!!!  That's fantastic. :-)   Ironically I don't use my own image, I use the "official" one.  Perhaps will need to take mine down and somehow point to the official one... or perhaps a new release of the latter might incentivize people?  @jpmens 

Anyway...  Docker images aren't really meant to be "extendable" (or even "configurable") once they are built; typically you fork and modify an existing `Dockerfile` and publish the resulting image yourself.  (Which isn't hard, especially now that Docker Hub has a builder.)

But thoughts about other ways to accomplish _some_ customization... 
- if it's just a matter of including `.py` files in some path for Python to find (I'm not a Python dev, so I don't know) then it would be easy enough to volume-mount an external folder for that purpose, as part of the image; by default it would be empty
- if modules or other programs need to be installed... you can start the image with an alternate entry point (i.e. bypass the Python command on the last line of the file), namely a shell, where you can do all your extra installation and configuration, and _then_ run Python. 

With the  latter, there might be a way to stop and restart that particular container, so as not to lose all your changes.  Hmm.

The thing to remember is that a `Dockerfile` is run only on _build_ -- once you have the image, the commands in the `Dockerfile` have long ago been executed.  I still get confused by that sometimes, thinking that the `Dockerfile` script is executed at runtime...

**Comment by amotl** [comment]
Dear @tinuva, @rgitzel, @tomhoover, @egadgetjnr and all Docker lovers,

#127 and #336 have come pretty far that we might think about improving the Dockerization once more.

> That said, I will be perfectly happy with a base mqttwarn image that is able to install additional dependencies on demand.

Service plugins can now be installed selectively by using `pip`, e.g. `pip install mqttwarn[xmpp,fbchat]`.

With kind regards,
Andreas.

**Comment by amotl** [comment]
### Just for the records...

@rgitzel asked on Oct 30, 2018:
> What packages do you install?

@tinuva has a nice [Dockerfile](https://github.com/tinuva/mqttwarn/blob/049e1539c7de3898e3cc4bbf00b87ecf5c41b6a0/Dockerfile) with some extended dependencies for serving more service plugins out of the box.

**Comment by amotl** [comment]
Hi again,

I believe we resolved this per #506. `mqttwarn-full` Docker images [1] will now be available for each upcoming release. [2] has corresponding documentation.

Please let us know about your experiences with those images and feel free to reopen this issue if you feel some details would need more attention.

Keep up the spirit and with kind regards,
Andreas.

[1] https://github.com/users/jpmens/packages/container/package/mqttwarn-full
[2] https://github.com/jpmens/mqttwarn/blob/0.23.1/DOCKER.md
